### ISSUE TITLE
Use case: representing Objective-C types and protocols

### ISSUE THREAD

**Issue by mjbshaw** [Centril/rfc-trait-parametric-polymorphism#12]
I'm working on [objrs](https://gitlab.com/objrs/objrs), a (mostly macro) library that transforms Rust code so it's compatible with Objective-C.

Objective-C protocols are one of the biggest hurdles for objrs. They're most analogous to Rust's traits. Unfortunately, there are some areas of Rust that make using a trait to represent an Objective-C protocol impossible.

For example, consider the Objective-C protocol [`NSCopying`](https://developer.apple.com/documentation/foundation/nscopying?language=objc). In Objective-C, you can use the type `id<NSCopying>` to reference any type that implements the `NSCopying` protocol. In Rust, the closest thing to this is something like this:

```rust
extern "C" {
  type Opaque;
}

struct Id<T: ?Sized>(core::marker::PhantomData<T>, Opaque);

trait NSCopying {
  // implementation details omitted for brevity
}

impl<T: NSCopying + ?Sized> NSCopying for Id<T> {}
```

Then, in my dream world, you could use the type `Id<NSCopying>` to represent any Objective-C type that implemented the `NSCopying` trait/protocol (of course, there's a lot more boilerplate to do this, but that's what objrs's macros are for).

Unfortunately, this doesn't work due to [RFC 255](https://github.com/rust-lang/rfcs/blob/master/text/0255-object-safety.md) and object safety. The `NSCopying` trait isn't object safe, and thus it can't be used as a trait object. To work around this, I've had to use a hack like this:

```rust
trait NSCopying {
  // implementation details omitted for brevity
}
struct NSCopyingId(Opaque);

impl NSCopying for NSCopyingId {}
impl<T: NSCopying + ?Sized> NSCopying for Id<T> {}
```

Now there are two versions of `NSCopying`: the trait (named `NSCopying`) that is to be used in things like generic trait bounds, and the concrete `Id`-compatible type (named `NSCopyingId`) that can be used when you need a type. This is unfortunate because now there are multiple names and multiple representations for a single Objective-C protocol (and it can lead to weird cases where someone has an `NSCopyingId` and someone else uses `Id<NSCopyingId>` and you need to get them to interop).

Trait-parametric polymorphism is *exactly* what I've been longing for to address this. It will make the Rust code clearer because it will allow a single name to be used, and I can use a consistent representation in Rust (a trait) rather than creating convoluted mixes of traits and types (which represent the same thing). With this RFC, I could actually pass the trait itself to `Id` and bypass the need for a trait object or extra struct type. It would make the code easier for users to read and understand.

**Comment by burdges** [comment]
Isn't Objective-C using basically what Rust calls "trait objects" everywhere under the hood?  I'd think `NSCopying` should just be designed to only work as `Box<dyn NSCopying>`, including `-> Box<dyn NSCopying>` methods, or maybe some `ObjCBox<dyn NSCopying>` type.

**Comment by mjbshaw** [comment]
@burdges That's exactly what I'm trying to do, but it doesn't work in certain situations ([like if the trait has a generic fn](https://rust.godbolt.org/z/aPw9JW), [or if the protocol has a class method](https://rust.godbolt.org/z/EHetwM)). I'm hoping that this RFC would at least provide a viable alternative.

**Comment by burdges** [comment]
Those type parameters must be type erased too.  As an example, check out [`serde::Serialize`](https://docs.serde.rs/serde/trait.Serialize.html) vs [`erased_serde::Serialize`](https://docs.serde.rs/erased_serde/trait.Serialize.html)

I'd think class methods would often have their "receiver" by some other type/trait, so you should make this implicit builder patterns explicit by making them methods of that other type/trait.  If they really have no receiver, and hence no mechanism to recognize the class, then you can explicitly exclude them from the trait object with a `Self: Sized` bound, like 
```
pub trait Foo {
    fn new() -> Self where Self: Sized
}
```
You also might explore normal and erased trait variants, like serde and serde_erased, but omit the class methods form the erased variant, or otherwise alter presentation.  If the class methods if identical for all classes then you can give trait objects inherent methods: 
```
impl dyn Trait {
    pub fn foo() { .. }
}
```

**Comment by mjbshaw** [comment]
Yeah, I've explored `where Self: Sized`, but that means the methods can't be `impl`ed for extern types (which are unsized). Extern types are too valuable to me in this case to give up. I could work around this by creating a separate sized type, but that results in the same poor API that I'm trying to avoid in the first place (and is the same kind of problem that splitting the trait into object-safe and non-object-safe variants hits: you've got two different types/traits for a single Objective-C protocol).
